src/violations.rs

```diff
@@ -3,52 +3,72 @@ use std::{
     hash::{Hash, Hasher},
 };
 
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub enum Violation {
+    DisallowedImportViolation(DisallowedImportViolation),
+    ReferenceToNonexistentDirectory(ReferenceToNonexistentDirectory),
+}
+
 #[derive(Debug)]
-pub struct Violation {
+pub struct DisallowedImportViolation {
     pub file_path: String,
     pub disallowed_import: String,
     pub full_disallowed_import: String,
 }
 
-impl Hash for Violation {
+impl Hash for DisallowedImportViolation {
     fn hash<H: Hasher>(&self, state: &mut H) {
         self.file_path.hash(state);
         self.disallowed_import.hash(state);
         self.full_disallowed_import.hash(state);
     }
 }
 
-impl PartialEq for Violation {
+impl PartialEq for DisallowedImportViolation {
     fn eq(&self, other: &Self) -> bool {
         self.file_path == other.file_path
             && self.disallowed_import == other.disallowed_import
             && self.full_disallowed_import == other.full_disallowed_import
     }
 }
 
-impl Eq for Violation {}
+impl Eq for DisallowedImportViolation {}
+
+#[derive(Debug, Eq, Hash, PartialEq)]
+pub struct ReferenceToNonexistentDirectory {
+    pub rules_file_path: String,
+    pub directory_name: String,
+    pub user_message: String,
```

Comment:
Unsure about what would be most idiomatic. Do not mind personally either way. 

This seemed a straightforward way to deal with both cases (missing in source position, or missing in target position) with a single struct.

It could also be 2 structs, or 1 struct and an enum, or a struct with some optionals + more logic to generate user message.

`rules_file_path` and `directory_name` are both there for `fix`.

Regardless of the violation type, information about the _location_ of the violation is probably always needed/desireable, and would be beneficial to have for things like displaying user messages and performing fixes/updates.

```
location: user_message
```